### ISSUE TITLE
HTTP API

### ISSUE THREAD

**Issue by Changaco** [liberapay/liberapay.com#688]
We need a proper HTTP API so that external apps can interface with Liberapay, and for any deep integration we need to become an OAuth provider so that apps can access and/or modify a user's donations.

**Comment by aggsol** [comment]
Why would any external app interface with Liberapay? What usecase do you see for creator and donors?

**Comment by Changaco** [comment]
Some preliminary thoughts:

- Should the API be part of the main app? Or should it be a separate app?
- `https://api.liberapay.com/` or `https://liberapay.com/api`? The subdomain is probably better in the long run because it's easy to route it separately from the main app.
- Should we use [GraphQL](http://graphql.org/)? Here's an interesting comparison: [GraphQL vs. REST](https://dev-blog.apollodata.com/graphql-vs-rest-5d425123e34b).

**Comment by aggsol** [comment]
I vote for: 
* a separated app makes it easier to move around, isolate, throttle etc
* `https://api.liberapay.com`
* REST is well know, easy to test (`curl` and `jq` ir more than enough) and liberapay is not complex enough IMO to benefit from GraphQL

**Comment by Changaco** [comment]
https://github.com/postgraphql/postgraphql
https://github.com/begriffs/postgrest

**Comment by Bat41** [comment]
I tried GraphQL once, and the first problem I had is there is no official solution for authentication. You can hack something, but it seems difficult to limit authentication to some resources. And as @aggsol said, GraphQL would probably be too much for Liberapay anyway.

**Comment by timpritlove** [comment]
I want to voice my support for an API and I want to give an example what kind of software could integrate such an API: online banking software (e.g. MoneyMoney which provides an internal API itself <https://moneymoney-app.com/api/webbanking/>).

The goal would be to integrate Liberapay into banking apps as if they were a real bank. To the user, there is not a big difference:

- there is authenticated access to an account
- the account has a balance
- there are transactions going on (payouts and withdrawals)

So if LP could be treated like a bank account, it would automatically integrate into online banking software that is itself has an open API. An even more useful step would be to actually provide FINTS [1] compatibility which would make LP compatible with all online banking software in Germany.

However, a simple HTTP-based API that provides balance and uniquely identified transactions would be a good start.

[1] http://www.hbci-zka.de/english/

**Comment by SirCmpwn** [comment]
I'm looking for an API, too. I want to use LiberaPay to accept donations, but I also accept donations from Patreon and a self-hosted (also FOSS) platform. On the latter I have an aggregated summary of donations from both sources: https://drewdevault.com/donate/

I would use the API to pull that data and display it alongside the other sources here.

**Comment by Changaco** [comment]
@SirCmpwn If all you need is basic public information about your account you can get that from `https://liberapay.com/$userid/public.json`, e.g. <https://liberapay.com/Liberapay/public.json>.

**Comment by SirCmpwn** [comment]
Nice, that'll do it.

**Comment by jelv** [comment]
It would be cool if sites like The Ocean Cleanup could use the oauth for there fund widget like https://www.theoceancleanup.com/fund/. Step 1 set funding and step 2 register or connect.

**Comment by ColinFinck** [comment]
I would like to stress the importance of an authenticated API to get information about the balance and transactions.
The ReactOS Project has recently started accepting donations over Liberapay due to popular demand. As the treasurer of the German non-profit organization backing ReactOS, I have to book each and every donation going to the project. When PayPal donations started to create an overwhelming amount of work, I wrote a plugin for our bookkeeping software to automate the import: https://www.reactos.org/blogs/inside-reactos-deutschland-ev-donations-and-how-automate-them

Currently, booking the few Liberapay donations is still possible manually, albeit tedious.
However, this puts us in a paradoxical situation: If Liberapay becomes too popular, we have to stop supporting it, because booking hundreds of donations manually isn't feasible. An API to automate this task could instantly resolve the situation.

**Comment by Changaco** [comment]
Technical note: representing money amounts as integers in JSON is not a good idea (I'm looking at you, Stripe), because it can easily result in an end-user being billed 100 times more (or less) than intended. We're currently representing money amounts as strings in our undocumented endpoints, and we should continue to use strings.

**Comment by SirCmpwn** [comment]
No no no, don't represent them as floats, that's even more error prone. Representing money as cents and integers is industry standard.

**Comment by Changaco** [comment]
@SirCmpwn I disagree, my personal experience is that integers are much more problematic. In python dealing with money amounts represented as strings is a piece of cake: `Decimal(amount)` when parsing and `str(amount)` when serializing. With integers the parsing and serializing involve figuring out whether or not the amount needs to be multiplied/divided by 100, which is very error prone and isn't guaranteed to be the same for all APIs (because each one may have a different list of "zero-decimal" currencies).

**Comment by SirCmpwn** [comment]
Not all languages have good decimal support like that. Consider JavaScript - where it's harder to fuck up integers than floats. If you always work in cents you're far more likely to *under*charge than *over*charge.

I mean, it's a matter of personal opinion for you. But it's a matter of industry standard for everyone else. Don't let your personal judgement make your API surprising and error-prone.

**Comment by Changaco** [comment]
Accidental floating point math errors don't result in numbers that are off by a factor of one hundred, they're far less dangerous than integers. Making an API more difficult to use from Python in order to make it easier to (mis)use from JavaScript seems like a bad trade-off to me.

> If you always work in cents

They're not always cents, some currencies are "[zero-decimal](https://stripe.com/docs/currencies#zero-decimal)".

**Comment by SirCmpwn** [comment]
Well, cents are a US dollar thing. There is no demonination smaller than 1 Yen. The idea is still basically the same, it's storing integers at a precision specific enough to include the smallest demonination.

**Comment by Changaco** [comment]
It's not always clear what the "smallest denomination" of a currency is, and it can change over time.

By the way, PayPal's REST API uses strings for money amounts, not integers.

**Comment by SirCmpwn** [comment]
Well, I've made my point. I'm not convinced but it's your API.

**Comment by danhunsaker** [comment]
A decimal would work fine as long as the payload _also_ includes the power of ten to adjust it by. US$ would, for example, have a value of 2 passed along with the amount, and the value can then be calculated internally as `amount / (10 ^ places)`. Zero-decimal currencies would pass 0 for the adjusted places, and since `10 ^ 0 == 1`, everything works as expected.

Strings, however, have this information built in to a single value. Language-specific precision issues can be bypassed by grabbing the integer form of the amount _with the decimal point already removed_ (it's generally fairly simple to remove all instances of a given character from a string), and counting the number of characters after the decimal point in the string value to determine the places adjustment.

Really, it's all sixes.

**Comment by danhunsaker** [comment]
Having chimed in on the technical side, I have another use case to propose.

Patreon permits OAuth logins to be used to unlock backer-only content on external sites. Most often, from what I've seen, this amounts to early access to things the general public will get to see later. My specific use case is exactly that – offering early access to web comic content for folks who back or otherwise donate, as a "thank you" rather than a purchase. I'd love to integrate Liberapay the same way, so that creators can enjoy this functionality without needing Patreon at all (or possibly even alongside it).

**Comment by Changaco** [comment]
@danhunsaker That use case is forbidden by the current terms of service. Donors must not receive any reward: no early access, no publicity, nothing. The discussion about modifying the terms of service is <https://github.com/liberapay/liberapay.org/issues/12>, I've copied your comment there.

**Comment by danhunsaker** [comment]
That's an entirely fair point. FLO as an ideology really doesn't support that use case, after all.

Luckily, other use cases presented here are still valid!

**Comment by decentral1se** [comment]
Hi folks, I'd like to chime in here and show a use case that is taking off in the world of Javascript. The open collective install banner: https://github.com/nuxt/opencollective. When you install a package, it shows the creator's project page and some meta data etc. I want to create this for liberapay but I'd need API access to get started.

**Comment by decentral1se** [comment]
btw, if there is some way forward, I wouldn't mind putting time into the implementation. Perhaps a first step to show some minimial metadata for the profile. That would probably cover my needs right now anyway ...

**Comment by alex-tee** [comment]
Hi, I have a little use case for an API too. I am adding the amounts from various income sources and I am displaying the amount on my website. For SendOwl and PayPal, I use their transaction APIs to get the amounts, but for LiberaPay I'm currently scraping it off the liberapay profile. Not a super popular use case but an API would be useful here too.

**Comment by nektro** [comment]
Being able to add a webhook for successful transactions would allow external count tracking for uses such as donor-rewards for reaching milestones, etc.

**Comment by 12people** [comment]
@Changaco Just curious — is this being worked on? If so, how far along is it?

**Comment by Changaco** [comment]
@12people I haven't started implementing this.

**Comment by 12people** [comment]
Ok, thanks for the info!

**Comment by MalteKiefer** [comment]
Is there any new status here?

**Comment by Changaco** [comment]
No. Having a proper HTTP API would be nice, but it's still not a priority.

I encourage you to share why you want an API, what you would like to do with it.

**Comment by MalteKiefer** [comment]
That is easy.
I want to create an Android App for the F-Store

Holen Sie sich Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Changaco** [comment]
Thank you for the clarification. An Android app has been discussed before in <https://github.com/liberapay/salon/issues/214>. I still think that a native app for Liberapay is too much work, and that working on #1000 is more efficient and effective.

**Comment by kairusds** [comment]
> No. Having a proper HTTP API would be nice, but it still isn't a priority.
> 
> I encourage you to share why you want an API, what you would like to do with it.

Having a proper API would be great if you want to get a list of the supporters/donators & your revenue

**Comment by Changaco** [comment]
If you're only looking for basic public information, you can find it in `https://liberapay.com/<username>/patrons/public.csv` and `https://liberapay.com/<username>/public.json`.

If you need the full private list of patrons, you can find the links in [your Patrons page](https://liberapay.com/about/me/patrons/), but it's difficult to automate the downloading of these private CSV files, because Liberapay currently only supports authentication tokens which eventually expire.

**Comment by brunoais** [comment]
> authentication tokens which eventually expire.

As long as there's a way to get a new authentication token, that's not a problem. How can a new token be obtained?

**Comment by kastwey** [comment]
Hello everyone.

I am the administrator of a Mastodon instance (mastodon.cat). I wanted to integrate a bot that we have, to send messages when we receive donations, but I see that you still don't have an API that allows this.
I have seen that there are endpoints to get public data, but I imagine that this does not include donations that people make privately, right?
Is there a way for me to know when I get a donation, whatever it is, a webhook maybe?

Thanks!

**Comment by Changaco** [comment]
@kastwey There is currently no way to be notified when you gain a new donor. If you want to know when you receive a payment, you could use [Stripe's webhooks](https://stripe.com/docs/webhooks), however you won't be able to distinguish initial payments from renewal payments.

You can easily get the list of your public donors, but currently there is no clean way to automate getting the list of your private donors.

**Comment by brunoais** [comment]
@Changaco How much work would be needed to allow automation to get the list of private donors?

**Comment by Changaco** [comment]
@brunoais Not much. The only thing missing is read-only authentication tokens which never expire.

However, since rewarding patrons is forbidden, the list of private donors has very few valid use cases.

**Comment by brunoais** [comment]
> However, since rewarding patrons is forbidden,

Took me a while to find it but you mean [this](https://liberapay.com/about/faq#differences), right?
![image](https://user-images.githubusercontent.com/639467/221543227-bbde1657-064c-4c96-a0b0-65ef2fa57abf.png)

I failed to spot that earlier.

**Comment by Changaco** [comment]
@brunoais You can share your use case for rewards in <https://github.com/liberapay/liberapay.org/issues/12>.

**Comment by osapon** [comment]
It would be nice to just have a link to download the CSV with a bearer token, etc.

Now, I am running a process that scrapes the login process, downloads the CSV, and imports it into my system, but sometimes I can't log in, as if some kind of scraping check is working.

**Comment by Changaco** [comment]
@osapon Your automated log-in attempts are probably hitting the new checks I added last week in response to a wave of abuse. I've started working on a proper mechanism for indefinite read-only access to an account. I should be able to land it before the end of next week since it's a relatively small change.

**Comment by Changaco** [comment]
It's now possible to obtain stable credentials for use in scripts: https://liberapay.com/about/me/access/constant-session

**Comment by osapon** [comment]
@Changaco Excellent!!
I'm glad I consulted. Thank you very much.

**Comment by tuktukyouy** [comment]
Hello,
Could we get public data from
https://liberapay.com/<username>/ledger

to something like 
https://liberapay.com/<username>/publicLedger.json

Like a list of all public donations username made?

**Comment by Changaco** [comment]
The list of public donors is in `https://liberapay.com/<username>/patrons/public.csv`.

**Comment by tuktukyouy** [comment]
> The list of public donors is in `https://liberapay.com/<username>/patrons/public.csv`.

Maybe I didnt made it clear sorry. I want the opposite, list of all my donations transactions. So something like in the ledger page.

**Comment by tuktukyouy** [comment]
Thanks for hard work, and If a donation goal is needed for your time, I could donate 100$ for the HTTP API option.

**Comment by JoeJoeTV** [comment]
> > The list of public donors is in `https://liberapay.com/<username>/patrons/public.csv`.
> 
> Maybe I didnt made it clear sorry. I want the opposite, list of all my donations transactions. So something like in the ledger page.

I am looking for the same thing, so that I can import this data into other money management software. Ideal would be info on all my current recurring donations (to whom, what amount of money and when it is to be paid)

**Comment by MathMan05** [comment]
I see it mentioned here, is OAuth planned? I'd like to add OAuth support into Harmony so they can connect their librepay to their Harmony accounts. (just for a link or to automate giving a donator badge to people maybe if that's allowed)
I just need it for authentication that users are who they say they are, so if there's an existing mechanism I could use instead let me know
